pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,321 @@
+		Name: name,
+		IPs:  ips,
+	}
+}
+
+func NewNodeLatencyMonitor(nodeInformer coreinformers.NodeInformer, interval, timeout time.Duration, limit int) *MonitorTool {
+	return &MonitorTool{
+		latencyStore: NewLatencyStore(nodeInformer),
+		interval:     interval,
+		timeout:      timeout,
+		limit:        limit,
+	}
+}
+
+func getRate(count int, d time.Duration, burst int) *rate.Limiter {
+	r := float64(count) / (d.Seconds() * 0.9)
+	return rate.NewLimiter(rate.Limit(r), burst)
+}
+
+func (m *MonitorTool) pingAll() {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+
+	// TODO: Get current node internal/external IP.
+	fromIP := ""
+
+	// A simple rate limiter
+	limiter := getRate(len(nodeIPs), m.interval, m.limit)
+	limitGroup := make(chan bool, int(m.limit))
+
+	klog.InfoS("Start to ping all nodes")
+	wg := sync.WaitGroup{}
+	for toIP, name := range nodeIPs {
+		limitGroup <- true
+		limiter.Wait(context.Background())
+		wg.Add(1)
+
+		go func(toIP string, name string) {
+			ctx, cancel := context.WithTimeout(context.Background(), m.timeout)
+			defer cancel()
+			defer wg.Done()
+
+			ok, rtt := m.pingNode(ctx, toIP)
+			if ok {
+				m.latencyStore.UpdateConnByKey(name, &Connection{
+					FromIP:      fromIP,
+					ToIP:        toIP,
+					Latency:     rtt,
+					Status:      true,
+					LastUpdated: time.Now(),
+				})
+			} else {
+				m.latencyStore.UpdateConnByKey(name, &Connection{
+					FromIP:      fromIP,
+					ToIP:        toIP,
+					Latency:     0,
+					Status:      false,
+					LastUpdated: time.Now(),
+				})
+			}
+
+			<-limitGroup
+
+		}(toIP, name)
+	}
+
+	wg.Wait()
+
+	// TODO: Print all connection status for debug
+	klog.InfoS("Finish to ping all nodes")
+	conns := m.latencyStore.ListConns()
+	for _, conn := range conns {
+		klog.InfoS("Connection status", "Connection", conn)
+	}
+}
+
+// resolveIP resolves the target address to an IP address.
+func resolveIPV4(ctx context.Context, target string) (*net.IPAddr, error) {
+	IPProtocol := "ip4"
+
+	klog.InfoS("Resolving target address", "ip_protocol", IPProtocol)
+
+	resolver := &net.Resolver{}
+	ips, err := resolver.LookupIPAddr(ctx, target)
+	if err != nil {
+		klog.ErrorS(err, "Resolution with IP protocol failed")
+		return nil, err
+	}
+
+	// Return the first IPv4 address
+	for _, ip := range ips {
+		if ip.IP.To4() != nil {
+			klog.InfoS("Resolved target address", "ip", ip.String())
+			return &ip, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find IPv4 address for target")
+}
+
+func (m *MonitorTool) pingNode(ctx context.Context, addr string) (bool, time.Duration) {
+	// Resolve the IP address
+	ip, err := resolveIPV4(ctx, addr)
+	if err != nil {
+		klog.ErrorS(err, "Failed to resolve IP address")
+
+		return false, 0
+	}
+
+	srcIP := net.ParseIP("0.0.0.0")
+	requestType := ipv4.ICMPTypeEcho
+	replyType := ipv4.ICMPTypeEchoReply
+
+	// TODO: Enable DontFragment?
+	socket, err := net.ListenPacket("ip4:icmp", srcIP.String())
```

Comment:
can we reuse the same socket instead of creating a new one every time?
and a more interesting question: can the same socket be used to "ping" multiple Nodes concurrently, as long as a single goroutine is responsible for all "writes" and another single goroutine is responsible for all "reads"? If it is possible, it would be a very good approach IMO.

Reply:
Sound great! I will try it.

Reply:
I've evaluated this and think it's somewhat similar to the pool concept, which hasn't been implemented in the current version for time reasons.

Reply:
We have splited ping to `sendPing` and `recvPing`, and share one socket to send requests.